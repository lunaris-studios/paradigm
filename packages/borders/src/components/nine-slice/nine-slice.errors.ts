import * as Props from "./nine-slice.props";

function error(message: string): string {
	return `${Props.DISPLAYNAME} | ${message}`;
}

export const WARN_NO_HEIGHT = error("height prop will default to 100px");
export const WARN_NO_CORNER = error("cornerLength prop will default to 25px");
export const WARN_NO_WIDTH = error("width prop will default to 100px");

export const ERROR_NO_IMAGE = error(
	"component failed to render, no image was provided for the 9-slice",
);
