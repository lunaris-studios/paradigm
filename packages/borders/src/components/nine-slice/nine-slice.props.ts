import * as Common from "~/common";

export const DISPLAYNAME = `${Common.DISPLAYNAME_PREFIX}.NineSlice`;
