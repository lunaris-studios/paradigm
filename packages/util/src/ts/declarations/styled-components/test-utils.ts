declare module "styled-components/test-utils" {
	import * as SC from "styled-components";

	type StyledComponent = SC.StyledComponent<any, any, any, any>;

	export function find(wrapper: Object, styledComponent: StyledComponent): Element;

	export function find(element: Element, styledComponent: StyledComponent): Element;

	export function findAll(element: Element, styledComponent: StyledComponent): NodeList;
}
