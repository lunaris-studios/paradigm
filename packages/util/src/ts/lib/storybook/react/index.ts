import * as Addons from "@storybook/addons";
import * as React from "react";

type ReactComponent = React.ComponentType<any> | React.FunctionComponent<any>;
type ReactReturnType = React.ReactElement<unknown>;

/**
 * Metadata to configure the stories for a component.
 *
 * @see [Default export](https://storybook.js.org/docs/formats/component-story-format/#default-export)
 */
export type Meta<Args = Addons.Args> = Addons.BaseMeta<ReactComponent> &
	Addons.Annotations<Args, ReactReturnType>;

/**
 * Story function that represents a component example.
 *
 * @see [Named Story exports](https://storybook.js.org/docs/formats/component-story-format/#named-story-exports)
 */
export type Story<Args = Addons.Args> = Addons.BaseStory<Args, ReactReturnType> &
	Addons.Annotations<Args, ReactReturnType>;
