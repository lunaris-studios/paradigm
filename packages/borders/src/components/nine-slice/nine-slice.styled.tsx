import * as Protocol from "@paradigmjs/protocol";
import * as SC from "styled-components";
import * as Spring from "react-spring";
// re-import `styled-components` development mode DOM classnames.
import styled, { css } from "styled-components";

import * as Util from "~/util";

import * as Component from "./nine-slice";
import * as Types from "./nine-slice.types";

/**
 * [NineSlice]
 * - [NineSlice.Container(WRAPPER)]
 * - - [NineSlice.Row(WRAPPER)(MAP<3>)]
 * - - - [NineSlice.Section(ELEMENT)(MAP<3>)]
 */

interface NineSlice {
	Container: Util.SC.Styled<Spring.AnimatedComponent<"div">, INineSliceContainerProps>;
	Row: Util.SC.Styled<"div", INineSliceRowProps>;

	StubImage: Util.SC.Styled<"img", INineSliceStubImageProps>;
	Section: Util.SC.Styled<"div", INineSliceSectionProps>;
}

export const NineSlice = {} as NineSlice;

/**
 * [NineSlice.Container]
 */

interface INineSliceContainerProps {
	corner: Component.INineSliceProps["corner"];
	height: Component.INineSliceProps["height"];
	width: Component.INineSliceProps["width"];
}

NineSlice.Container = styled(Spring.animated.div)<INineSliceContainerProps>`
	${(props: INineSliceContainerProps) => {
		const { height, width, corner } = props;

		const heightPx = Protocol.Snippets.value(height, Protocol.Unit.PX);
		const widthPx = Protocol.Snippets.value(width, Protocol.Unit.PX);
		const cornerPX = Protocol.Snippets.value(corner, Protocol.Unit.PX);

		return css`
			height: ${heightPx};
			width: ${widthPx};
			grid-template-columns: ${cornerPX} 1fr ${cornerPX};
			grid-template-rows: ${cornerPX} 1fr ${cornerPX};
		`;
	}}
	/** */
	display: grid;
	grid-template-areas:
		"${Types.NineSliceCoordinate.NORTH_WEST} ${Types.NineSliceCoordinate.NORTH} ${
	Types.NineSliceCoordinate.NORTH_EAST
}"
		"${Types.NineSliceCoordinate.WEST} ${Types.NineSliceCoordinate.CENTER} ${
	Types.NineSliceCoordinate.EAST
}"
		"${Types.NineSliceCoordinate.SOUTH_WEST} ${Types.NineSliceCoordinate.SOUTH} ${
	Types.NineSliceCoordinate.SOUTH_EAST
}";
`;

/**
 * [NineSlice.Row]
 */

interface INineSliceRowProps {}

NineSlice.Row = styled("div")<INineSliceRowProps>``;

/**
 * [NineSlice.StubImage]
 */

interface INineSliceStubImageProps {}

NineSlice.StubImage = styled("img")<INineSliceStubImageProps>`
	display: none;
`;

/**
 * [NineSlice.Section]
 */

interface INineSliceSectionProps {
	coordinate: Types.NineSliceCoordinate;
	corner: Component.INineSliceProps["corner"];
	height: Component.INineSliceProps["height"];
	image: Component.INineSliceProps["image"];
	imageSize: Component.INineSliceState["imageSize"];
	width: Component.INineSliceProps["width"];
}

NineSlice.Section = styled("div")<INineSliceSectionProps>`
	/* ${Protocol.Snippets.debug()} */
	/** */
	${(props: INineSliceSectionProps) => {
		const { coordinate, corner, height, image, imageSize, width } = props;

		const bgImage = Boolean(coordinate !== Types.NineSliceCoordinate.CENTER)
			? `url(${image})`
			: "";

		const bgPosition = getSectionBgPosition(coordinate);

		const bgSize = getSectionBgSize(coordinate, corner, imageSize, height, width);
		const bgSizeWidth = Boolean(bgSize.width)
			? Protocol.Snippets.value(bgSize.width, Protocol.Unit.PX)
			: "auto";
		const bgSizeHeight = Boolean(bgSize.height)
			? Protocol.Snippets.value(bgSize.height, Protocol.Unit.PX)
			: "auto";

		const display = Boolean(coordinate === Types.NineSliceCoordinate.CENTER)
			? "contents"
			: "block";

		return css`
			background-image: ${bgImage};
			background-position-x: ${bgPosition.x};
			background-position-y: ${bgPosition.y};
			background-size: ${bgSizeWidth} ${bgSizeHeight};
			display: ${display};
			grid-area: ${coordinate};
		`;
	}}
`;

function getSectionBgSize(
	coordinate: Types.NineSliceCoordinate,
	corner: INineSliceSectionProps["corner"],
	imageSize: INineSliceSectionProps["imageSize"],
	height: INineSliceSectionProps["height"],
	width: INineSliceSectionProps["width"],
): Types.INineSliceSectionBgSize {
	const size: Types.INineSliceSectionBgSize = {
		height: 0,
		width: 0,
	};

	const widthScaleOffset = Math.ceil((corner / imageSize!.width) * width);
	const heightScaleOffset = Math.ceil((corner / imageSize!.height) * height);

	switch (coordinate) {
		/** Top & Bottom */
		case Types.NineSliceCoordinate.NORTH:
		case Types.NineSliceCoordinate.SOUTH:
			size.height = imageSize!.height;
			size.width = width + widthScaleOffset;
			break;

		/** Left & Right */
		case Types.NineSliceCoordinate.EAST:
		case Types.NineSliceCoordinate.WEST:
			size.height = height + heightScaleOffset;
			size.width = imageSize!.width;
			break;

		case Types.NineSliceCoordinate.CENTER:
			size.height = height - corner * 2;
			size.width = width - corner * 2;
	}

	return size;
}

function getSectionBgPosition(
	coordinate: Types.NineSliceCoordinate,
): Types.INineSliceSectionBackgroundImageOffset {
	const backgroundPosition: Types.INineSliceSectionBackgroundImageOffset = {
		x: "center",
		y: "center",
	};

	switch (coordinate) {
		case Types.NineSliceCoordinate.NORTH:
			backgroundPosition.x = "center";
			backgroundPosition.y = "top";
			break;

		case Types.NineSliceCoordinate.SOUTH:
			backgroundPosition.x = "center";
			backgroundPosition.y = "bottom";
			break;

		case Types.NineSliceCoordinate.EAST:
			backgroundPosition.x = "right";
			backgroundPosition.y = "center";
			break;

		case Types.NineSliceCoordinate.WEST:
			backgroundPosition.x = "left";
			backgroundPosition.y = "center";
			break;

		case Types.NineSliceCoordinate.NORTH_WEST:
			backgroundPosition.x = "left";
			backgroundPosition.y = "top";
			break;

		case Types.NineSliceCoordinate.NORTH_EAST:
			backgroundPosition.x = "right";
			backgroundPosition.y = "top";
			break;

		case Types.NineSliceCoordinate.SOUTH_EAST:
			backgroundPosition.x = "right";
			backgroundPosition.y = "bottom";
			break;

		case Types.NineSliceCoordinate.SOUTH_WEST:
			backgroundPosition.x = "left";
			backgroundPosition.y = "bottom";
			break;

		default:
			break;
	}

	return backgroundPosition;
}
