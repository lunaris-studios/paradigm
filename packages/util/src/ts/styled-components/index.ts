import * as SC from "styled-components";

/**
 * Condensed typing for a `styled-component`.
 */
export type Styled<
	TElement extends keyof JSX.IntrinsicElements | React.ComponentType<any>,
	TProps extends object,
	TAttributes extends object = {}
> = SC.StyledComponent<
	TElement,
	SC.DefaultTheme,
	TProps & TAttributes,
	keyof TAttributes
>;
