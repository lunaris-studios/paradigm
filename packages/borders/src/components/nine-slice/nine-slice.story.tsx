import * as React from "react";

import * as Util from "~/util";

import * as Component from "./nine-slice";

import BORDER from "~/assets/border.png";
import SWORD from "~/assets/sword.png";

const Template: Util.Storybook.React.Story<Component.INineSliceProps> = (args) => (
	<Component.NineSlice {...args} image={BORDER} corner={22} />
);

export const Basic = Template.bind({});
Basic.args = {};

export const Vertical = Template.bind({});
Vertical.args = {
	height: 256,
	width: 128,
};

export const Horizontal = Template.bind({});
Horizontal.args = {
	height: 128,
	width: 256,
};

export const Big = Template.bind({});
Big.args = {
	height: 512,
	width: 512,
};

export const Content = Template.bind({});
Content.args = {
	children: (
		<img
			src={SWORD}
			style={{
				height: "100%",
				width: "100%",
				objectFit: "contain",
			}}
		/>
	),
};

export default {
	title: "Borders/NineSlice",
	component: Component.NineSlice,
} as Util.Storybook.React.Meta;
