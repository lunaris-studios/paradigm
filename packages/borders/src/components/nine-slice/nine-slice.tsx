import * as Abstract from "@paradigmjs/abstract";
import * as React from "react";
import * as Spring from "react-spring";
import styled, { css } from "styled-components";

import * as Common from "~/common";

import * as Errors from "./nine-slice.errors";
import * as Styled from "./nine-slice.styled";
import * as Types from "./nine-slice.types";

export interface INineSliceProps {
	/**
	 * Squared length reserved for the non-scalable, static corner elements.
	 * of the 9-slice layout.
	 * @default 8;
	 */
	corner: number;

	/**
	 * Passed as a child to the `Types.NineSliceCoordinate.CENTER` 9-slice section.
	 */
	children?: React.ReactNode;

	/**
	 * Height of the surface.
	 * @default 128
	 */
	height: number;

	/**
	 * File path of the image to use for the background of the 9-slice
	 * surface.
	 */
	image: string;

	/**
	 * Styling to apply to the child element.
	 */
	style?: object;

	/**
	 * HTML tag to use for element.
	 * @default "div"
	 */
	tagName: keyof JSX.IntrinsicElements;

	/**
	 * Width of the surface.
	 * @default 128
	 */
	width: number;
}

export interface INineSliceDefaultProps
	extends Omit<INineSliceProps, "image" | "children"> {}

export interface INineSliceState {
	/**
	 * Height & width of the 9-slice image.
	 * @default undefined
	 */
	imageSize: Types.INineSliceImageSize | undefined;

	/**
	 * Indicates whether or not the image has finished loading.
	 * @default false
	 */
	isLoaded: boolean;
}

const defaultProps = Object.freeze<INineSliceDefaultProps>({
	corner: 8,
	height: 128,
	tagName: "div",
	width: 128,
});

const defaultState = Object.freeze<INineSliceState>({
	imageSize: undefined,
	isLoaded: false,
});

export class NineSlice extends Abstract.AbstractPureComponent<
	INineSliceProps,
	INineSliceState
> {
	public static readonly displayName = `${Common.DISPLAYNAME_PREFIX}.NineSlice`;

	public constructor(props: INineSliceProps) {
		super(props);

		this.onImgLoad = this.onImgLoad.bind(this);
	}

	public static readonly defaultProps: INineSliceDefaultProps = defaultProps;
	public state: INineSliceState = defaultState;

	public static Styled = Styled;
	public static Errors = Errors;
	public static Types = Types;

	public render(): JSX.Element {
		const { children, corner, height, image, tagName, width } = this.props;
		const { imageSize, isLoaded } = this.state;

		const sectionProps = { corner, height, image, imageSize, width };

		return (
			<React.Fragment>
				<Styled.NineSlice.StubImage src={image} onLoad={this.onImgLoad} />
				<Spring.Transition items={isLoaded} from={{ opacity: 0 }} enter={{ opacity: 1 }}>
					{(springs, isLoaded) =>
						isLoaded && (
							<Styled.NineSlice.Container
								corner={corner}
								height={height}
								style={springs}
								width={width}
							>
								{/* Row One */}
								<Styled.NineSlice.Section
									{...sectionProps}
									coordinate={Types.NineSliceCoordinate.NORTH_WEST}
								/>
								<Styled.NineSlice.Section
									{...sectionProps}
									coordinate={Types.NineSliceCoordinate.NORTH}
								/>
								<Styled.NineSlice.Section
									{...sectionProps}
									coordinate={Types.NineSliceCoordinate.NORTH_EAST}
								/>

								{/* Row Two */}
								<Styled.NineSlice.Section
									{...sectionProps}
									coordinate={Types.NineSliceCoordinate.WEST}
								/>
								<Styled.NineSlice.Section
									{...sectionProps}
									coordinate={Types.NineSliceCoordinate.CENTER}
								>
									{children}
								</Styled.NineSlice.Section>
								<Styled.NineSlice.Section
									{...sectionProps}
									coordinate={Types.NineSliceCoordinate.EAST}
								/>

								{/* Row Three */}
								<Styled.NineSlice.Section
									{...sectionProps}
									coordinate={Types.NineSliceCoordinate.SOUTH_WEST}
								/>
								<Styled.NineSlice.Section
									{...sectionProps}
									coordinate={Types.NineSliceCoordinate.SOUTH}
								/>
								<Styled.NineSlice.Section
									{...sectionProps}
									coordinate={Types.NineSliceCoordinate.SOUTH_EAST}
								/>
							</Styled.NineSlice.Container>
						)
					}
				</Spring.Transition>
			</React.Fragment>
		);
	}

	/** */

	protected validateProps(props: INineSliceProps): void {
		const { height, image, corner, width } = props;

		if (height == null) {
			console.warn(Errors.WARN_NO_HEIGHT);
		}
		if (image == null) {
			console.error(Errors.ERROR_NO_IMAGE);
		}
		if (corner == null) {
			console.warn(Errors.WARN_NO_CORNER);
		}
		if (width == null) {
			console.warn(Errors.WARN_NO_WIDTH);
		}
	}

	public onImgLoad(event: React.SyntheticEvent<HTMLImageElement, Event>): void {
		const { currentTarget: img } = event;

		this.setState({
			imageSize: {
				height: img.naturalHeight,
				width: img.naturalWidth,
			},
			isLoaded: true,
		});
	}
}
