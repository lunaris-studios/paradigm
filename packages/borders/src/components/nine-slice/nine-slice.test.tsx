import * as Protocol from "@paradigmjs/protocol";
import * as React from "react";
import * as Testing from "@testing-library/react";
import * as SCTU from "styled-components/test-utils";

import * as Component from "./nine-slice";
import * as Styled from "./nine-slice.styled";

describe("Nine", () => {
	it("renders correctly", () => {
		const node = Testing.render(<Component.NineSlice image="stub" />);
		const NineSliceContainer = SCTU.find(node.baseElement, Styled.NineSlice.Container);

		expect(NineSliceContainer).toMatchSnapshot();
	});
});
