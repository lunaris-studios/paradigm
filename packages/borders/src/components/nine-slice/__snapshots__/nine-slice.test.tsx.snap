// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nine renders correctly 1`] = `null`;
